tests/hardware/test_em27_webscraper.py

```diff
@@ -0,0 +1,148 @@
+"""Tests for the EM27Scraper class."""
+from decimal import Decimal
+from importlib import resources
+from pathlib import Path
+from unittest.mock import MagicMock, Mock, patch
+
+import pytest
+from PySide6.QtNetwork import QNetworkReply
+
+from finesse.config import EM27_URL
+from finesse.hardware.em27_scraper import (
+    EM27Error,
+    EM27Property,
+    EM27Scraper,
+    get_em27sensor_data,
+)
+
+
+@pytest.fixture
+def em27_property():
+    """Fixture for EM27Property."""
+    return EM27Property("Voltage", Decimal(1.23), "V")
+
+
+def test_str(em27_property: EM27Property):
+    """Test EM27Property's __str__() method."""
+    assert em27_property.__str__() == "Voltage = 1.230000 V"
+
+
+def test_val_str(em27_property: EM27Property):
+    """Test EM27Property's val_str() method."""
+    assert em27_property.val_str() == "1.230000 V"
+
+
+@pytest.fixture
+def em27_scraper(qtbot) -> EM27Scraper:
+    """Fixture for EM27Scraper."""
+    return EM27Scraper()
+
+
+def test_init(subscribe_mock: MagicMock) -> None:
+    """Test EM27Scraper's constructor."""
+    scraper = EM27Scraper()
+    assert scraper._url == EM27_URL
+    subscribe_mock.assert_called_once_with(scraper.send_data, "em27.data.request")
+
+
+@patch("finesse.hardware.em27_scraper.get_em27sensor_data")
+def test_on_reply_received_no_error(
+    get_em27sensor_data_mock: Mock, em27_scraper: EM27Scraper, sendmsg_mock: Mock, qtbot
+) -> None:
```

Comment:
You can have multiple `patch` decorators, so you could leave `_on_reply_received` as a non-member function.

They are in the opposite order to the arguments though, i.e.:

```py
@patch("finesse.hardware.em27_scraper._on_reply_received")
@patch("finesse.hardware.em27_scraper.get_em27sensor_data")
def test_on_reply_received_no_error(
    get_em27sensor_data_mock: Mock, reply_received_mock: Mock, em27_scraper: EM27Scraper, sendmsg_mock: Mock, qtbot
) -> None:
```

Reply:
Marking as resolved as per Teams discussion - i.e. no need to patch _on_reply_received()